tools/buildFactory.ts

```diff
@@ -174,73 +202,166 @@ export async function gulpUmdMinFactory(input: string, output: string) {
     globs: string | string[],
     opts?: SrcOptions
   },
   output: string,
-  overrideSettings?: Omit<ts.Settings, 'outDir'>,
-  useCjsTransform = false
+  overrideSettings?: Omit<gulpTs.Settings, 'outDir'>
 ) {
-  const dependencySolver = new DependencySolver(
-    input.sourceRoots,
-    {
-      baseUrl: input.opts?.base ?? '.',
-      outDir: output
-    }
-  );
+  output = `dist/isolate/${output}`;
+  const indexFileDir = 'dist/isolate/.build-data';
 
-  function build() {
-    return gulpFactory(
-      input,
-      output,
-      {
-        ...overrideSettings,
-        getCustomTransformers: (program?: Program) => {
-          const treeShakerTransform = tsTreeShaker({
-            solver: dependencySolver,
-            indexFile: input.indexFile
-          });
-
-          const customTransformers = overrideSettings?.getCustomTransformers?.(program);
-          return {
-            before: customTransformers?.before,
-            after: [
-              ...treeShakerTransform.after,
-              ...customTransformers?.after ?? []
-            ],
-            afterDeclarations: [
-              ...treeShakerTransform.afterDeclarations,
-              ...customTransformers?.afterDeclarations ?? []
-            ]
-          };
-        }
-      },
-      useCjsTransform
-    );
+  async function generateIndex() {
+    // get files from glob
+    const globs = Array.isArray(input.globs) ? input.globs : [input.globs];
+    const files = globs
+      .map((pattern) => glob.sync(pattern, input.opts))
+      .reduce((acc, val) => acc.concat(val), []);
+    // generate sha256 hash of the source code
+    const hash = crypto.createHash('sha256');
+    hash.update(input.isolateBuildConfig.toString());
+    files.forEach((file) => hash.update(fs.readFileSync(file)));
+    const hashValue = hash.digest('hex');
+    // compare hash with the previous one
+    const previousHash = fs.existsSync(`${indexFileDir}/hash`)
+      ? fs.readFileSync(`${indexFileDir}/hash`, 'utf8')
+      : '';
+    if (previousHash === hashValue) return;
+    // generate index file
+    const project = gulpTs.createProject('tsconfig.json', {
+      target: 'ESNext',
+      declaration: true,
+      getCustomTransformers: (program) => {
+        if (program === undefined) throw new Error('program is undefined');
+        return exportTransformer(program, {
+          indexOutputPath: `${indexFileDir}/transformed-index.json`,
+          indexFile: input.indexFile,
+          builds: input.isolateBuildConfig.builds
+            .map((build) => ({ name: build.name, sourceRoots: [build.sourceRoot] }))
+        });
+      }
+    });
+    await new Promise((resolve, reject) => {
+      gulp.src(input.globs, input.opts)
+        .pipe(project())
+        .on('error', reject)
+        .on('end', resolve)
+        .on('finish', resolve);
+    });
+    // save hash
+    fs.mkdirSync(indexFileDir, { recursive: true });
+    fs.writeFileSync(`${indexFileDir}/hash`, hashValue);
```

Comment:
Maybe create a temp file?

Change the source code directly is a dangerous action.